### DISCUSSION TITLE
How should I upload files from a data url (webcam) with Livewire?

### DISCUSSION THREAD

**Discussion by ralphjsmit** [livewire/livewire#4095] [1. Help]
I'm using Livewire and I need to capture an image from the user's webcam.

I'm not very confident in Javascript, so I was happy to discover that Livewire also has helpers for file uploads (great!), besides all the other awesome stuff.

I found this [very explanatory article](https://www.html5rocks.com/en/tutorials/getusermedia/intro/) about capturing the webcam, displaying the video stream and making a screenshot when the user clicks a button. 

The uploaded tile is then returned as a data url (appended to the src of an image, it looks like this`src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAoAAAAHgCAYAAAA .....`).

How should I upload this to Livewire? I searched on Google, but it looks like I can't directly append this to an `<input type="file"`. Is that correct?

If it is indeed not possible, can I use the `@this` helper to upload the file?

```js
    // Upload a file:
    @this.upload('photo', file, (uploadedFilename) => {
        // Success callback.
    }, () => {
        // Error callback.
    }, (event) => {
        // Progress callback.
        // event.detail.progress contains a number between 1 and 100 as the upload progresses.
   })
```

Many thanks!

**Comment by akhan619** [comment]
I am no expert with webcams but can you get the screenshot as a blob. If so, then you can bind the blob object to a File and upload that. I recently had to do it for image cropping using Croppie. There I can get a blob of the cropped image which can be uploaded to livewire backend as follows: ( Note this is javascript that I was using with Alpine)

```javascript
(blob) => {	   
                        
  // Make a new cropped image file using the blob object
  let file = new File([blob], 'temporaryImageFile.png', {type:"image/png", lastModified:new Date().getTime()});
  
  // Upload the file:
  @this.upload('final_image', file, (uploadedFilename) => {
      // Success callback.
  }, () => {
      // Error callback.
  }, (event) => {
      // Progress callback.
      // event.detail.progress contains a number between 1 and 100 as the upload progresses.
});
```

Ofcourse, for this to work you should have a public property called ```final_image```. The good thing is that you can use Laravel file validation as usual. No need to work with base64 strings and all. I am not sure but maybe there is a way to convert base64 image to blob in javascript.

**Reply by joshhanley** [reply]
@ralphjsmit yeah I don't have any experience with webcams either, but try out the suggestion above from @akhan619 and let us know if you have any issues!

**Reply by ralphjsmit** [reply]
@akhan619 Thank you for thinking along and providing this code! I don't know if I can get the screenshot as a blob directly, but I see that I should be able to [convert a data url to a blob](https://stackoverflow.com/questions/12168909/blob-from-dataurl).

I'm planning to try it out this week and will post my experiences here!
